primitives/src/proposal.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Proposals for relay-chain governance.
+//!
+//! This describes a combination of a function ID and data that can be used to call into
+//! an internal function.
+
+use bytes::Vec;
+use block::Number as BlockNumber;
+use codec::Slicable;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[repr(u8)]
+enum InternalFunctionId {
+	/// Set the system's code.
+	SystemSetCode = 0,
+	/// Set the number of sessions per era.
+	StakingSetSessionsPerEra = 1,
+	/// Set the minimum bonding duration for staking.
+	StakingSetBondingDuration = 2,
+	/// Set the validator count for staking.
+	StakingSetValidatorCount = 3,
+	/// Set the per-mille of validator approval required for governance changes.
+	GovernanceSetApprovalPpmRequired = 4,
+	/// Set the session length.
+	SessionSetLength = 5,
+}
+
+impl InternalFunctionId {
+	/// Derive `Some` value from a `u8`, or `None` if it's invalid.
+	fn from_u8(value: u8) -> Option<InternalFunctionId> {
+		use self::*;
+		let functions = [
```

Comment:
clobbered my change to use a transformation means that doesn't require the sequence to be wholly contiguous?